lib/history/sqlite.go

```diff
@@ -0,0 +1,224 @@
+func NewSQLiteTimeline(database *sql.DB, size int) (Timeline, error) {
+	if err := initEventTable(database); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &SQLiteTimeline{
+		size:       size,
+		database:   database,
+		lastStatus: &Cluster{Status: pb.SystemStatus_Unknown.String()},
+	}, nil
+}
+
+// initEventTable initializes event table schema.
+func initEventTable(db *sql.DB) error {
+	if _, err := db.Exec(createTableEvent); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// RecordStatus records differences of the previous status to the provided
+// status into the Timeline.
+func (t *SQLiteTimeline) RecordStatus(ctx context.Context, status *pb.SystemStatus) error {
+	cluster := parseSystemStatus(status)
+	events := t.lastStatus.diffCluster(cluster)
+	if len(events) == 0 {
+		return nil
+	}
+
+	tx, err := t.database.BeginTx(ctx, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := t.insertEvents(events); err != nil {
+		tx.Rollback()
+		return trace.Wrap(err, "failed to insert events.")
+	}
+
+	if err := t.evictEvents(); err != nil {
+		tx.Rollback()
+		return trace.Wrap(err, "failed to evict old events.")
+	}
+
+	if err := tx.Commit(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	t.lastStatus = cluster
+	return nil
+}
+
+// GetEvents returns the current timeline.
+func (t *SQLiteTimeline) GetEvents() (events []*Event, err error) {
+	rows, err := t.database.Query(selectAllFromEvent)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var (
+		id        int
+		timestamp time.Time
+		eventType EventType
+		node      string
+		probe     string
+		old       string
+		new       string
+	)
+
+	for rows.Next() {
+		if err := rows.Scan(&id, &timestamp, &eventType, &node, &probe, &old, &new); err != nil {
+			rows.Close()
+			return nil, trace.Wrap(err)
+		}
+
+		var event *Event
+		switch eventType {
+		case ClusterRecovered:
+			event = NewClusterRecoveredEvent(timestamp, old, new)
+		case ClusterDegraded:
+			event = NewClusterDegradedEvent(timestamp, old, new)
+		case NodeAdded:
+			event = NewNodeAddedEvent(timestamp, node)
+		case NodeRemoved:
+			event = NewNodeRemovedEvent(timestamp, node)
+		case NodeRecovered:
+			event = NewNodeRecoveredEvent(timestamp, node, old, new)
+		case NodeDegraded:
+			event = NewNodeDegradedEvent(timestamp, node, old, new)
+		case ProbePassed:
+			event = NewProbePassedEvent(timestamp, node, probe, old, new)
+		case ProbeFailed:
+			event = NewProbeFailedEvent(timestamp, node, probe, old, new)
+		default:
+			event = NewUnknownEvent(timestamp)
+		}
+
+		events = append(events, event)
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return events, nil
+}
+
+// insertEvents inserts the provided events into the timeline.
+func (t *SQLiteTimeline) insertEvents(events []*Event) error {
+
+	// prepare bulk insert statement
+	valueStrings := make([]string, 0, len(events))
+	valueArgs := make([]interface{}, 0, len(events)*eventSize)
```

Comment:
nit: `eventSize` belongs inside this method at least.
Also, why do you have `node`, `probe`, `old` and `new` inside the metadata map as opposed to being individual fields? The way they're handled here makes me believe they're not optional (as other metadata might be)

Reply:
Will leave it as metadata for this PR, but will work on restructuring Events and protobuf messages in a separate PR.